cylc/flow/workflow_db_mgr.py

```diff
@@ -112,27 +115,27 @@ def __init__(self, pri_d=None, pub_d=None):
             self.TABLE_ABS_OUTPUTS: []}
         self.db_updates_map = {}
 
-    def copy_pri_to_pub(self):
-        """Copy content of primary database file to public database file.
-
-        Use temporary file to ensure that we do not end up with a partial file.
-
-        """
-        temp_pub_db_file_name = None
+    def copy_pri_to_pub(self) -> None:
+        """Copy content of primary database file to public database file."""
         self.pub_dao.close()
+        # Use temporary file to ensure that we do not end up with a
+        # partial file.
+        # If an external connection is locking the old public db, it will
+        # still be connected to its inode, but should no longer affect future
+        # accesses (hopefully that process will soon recover to give up
+        # the lock).
+        temp_pub_db_fd, temp_pub_db_file_name = mkstemp(
+            prefix=self.pub_dao.DB_FILE_BASE_NAME,
+            dir=os.path.dirname(self.pub_dao.db_file_name)
+        )
+        os.close(temp_pub_db_fd)
         try:
-            self.pub_dao.conn = None  # reset connection
-            open(self.pub_dao.db_file_name, "a").close()  # touch
-            st_mode = os.stat(self.pub_dao.db_file_name).st_mode
-            temp_pub_db_file_name = mkstemp(
-                prefix=self.pub_dao.DB_FILE_BASE_NAME,
-                dir=os.path.dirname(self.pub_dao.db_file_name))[1]
             copy(self.pri_dao.db_file_name, temp_pub_db_file_name)
             os.rename(temp_pub_db_file_name, self.pub_dao.db_file_name)
-            os.chmod(self.pub_dao.db_file_name, st_mode)
-        except (IOError, OSError):
-            if temp_pub_db_file_name:
-                os.unlink(temp_pub_db_file_name)
+            os.chmod(self.pub_dao.db_file_name, PERM_PUBLIC)
```

Comment:
Just occurred to me, could the public db permissions be different to 0644 for a pre-existing db?

Reply:
Looks like before we weren't setting the permission of public DB, so it came out with the default os mask.

I think it should be left that way, if the default mask is not good enough for public access then it doesn't matter what the permissions on the file are as the workflow run directory will not be visible anyway.

The user might have a non-standard mark depending on site-policy e.g. they might want `664`, `640`, whatever.